js_modules/dagster-ui/packages/ui-core/src/asset-data/AssetLiveDataThread.tsx

```diff
@@ -0,0 +1,255 @@
+import {ApolloClient, gql} from '@apollo/client';
+
+import {LiveDataForNode, tokenForAssetKey, buildLiveDataForNode} from '../asset-graph/Utils';
+import {AssetKeyInput} from '../graphql/types';
+
+import type {AssetLiveDataThreadManager} from './AssetLiveDataThreadManager';
+import {
+  AssetGraphLiveQuery,
+  AssetGraphLiveQueryVariables,
+  AssetNodeLiveFragment,
+} from './types/AssetLiveDataThread.types';
+import {BATCHING_INTERVAL} from './util';
+
+export type AssetLiveDataThreadID = 'default' | 'sidebar' | 'asset-graph';
+
+export class AssetLiveDataThread {
+  private isFetching: boolean = false;
+  private listenersCount: {[key: string]: number};
+  private client: ApolloClient<any>;
+  private isLooping: boolean = false;
+  private interval?: ReturnType<typeof setTimeout>;
+  private manager: AssetLiveDataThreadManager;
+  public pollRate: number = 30000;
+
+  protected static _threads: {[key: string]: AssetLiveDataThread} = {};
+
+  constructor(client: ApolloClient<any>, manager: AssetLiveDataThreadManager) {
+    this.client = client;
+    this.listenersCount = {};
+    this.manager = manager;
+  }
+
+  public setPollRate(pollRate: number) {
+    this.pollRate = pollRate;
+  }
+
+  public subscribe(key: string) {
+    this.listenersCount[key] = this.listenersCount[key] || 0;
+    this.listenersCount[key] += 1;
+    this.startFetchLoop();
+  }
+
+  public unsubscribe(key: string) {
+    this.listenersCount[key] -= 1;
+    if (this.listenersCount[key] === 0) {
+      delete this.listenersCount[key];
+    }
+    if (this.getObservedKeys().length === 0) {
+      this.stopFetchLoop();
+    }
+  }
+
+  public getObservedKeys() {
+    return Object.keys(this.listenersCount);
+  }
+
+  public startFetchLoop() {
+    if (this.isLooping) {
+      return;
+    }
+    this.isLooping = true;
+    const fetch = () => {
+      const nextAssets = this.manager.determineAssetsToFetch(this.getObservedKeys());
+      if (nextAssets.length) {
+        this._batchedQueryAssets(nextAssets);
+      }
+    };
+    setTimeout(fetch, BATCHING_INTERVAL);
+    this.interval = setInterval(fetch, 5000);
```

Comment:
Kind of odd that this only uses BATCHING_INTERVAL the first time?

edit: Actually sorta seems like maybe only one of these two lines should be here?

Reply:
the interval is a continuous loop, BATCHING_INTERVAL is only used the first time to ensure we start fetching immediately. BATCHING_INTERVAL ensures we wait for all synchronous subscriptions to occur before we fetch though (if it were done synchronously we would fetch only a single asset in the first fetch)